### ISSUE TITLE
Cannot join entities with custom join columns

### ISSUE THREAD

**Issue by epechuzal** [mikro-orm/mikro-orm#593]
**Describe the bug**

We have a database where the primary key (mysql auto-increment) is not actually the key which we use to refer to entities and make joins - for that, there is a `key` field on the entity which is just some string which looks like a UUID.

My problem is that even though I set the `joinColumn` attributes on both sides of the relationship, MikroORM ignores this and tries to use the `id` field on the `OneToMany` side

With example entities:
```ts
export class ParentEntity {
  @OneToMany({
    entity: () => ChildEntity,
    mappedBy: e => e.parentEntity,
    joinColumn: 'key'
  })
  children = new Collection<ChildEntity>(this);

  @PrimaryKey()
  id!: number;

  @Property({ length: 191 })
  key!: string;

  @Property({ length: 191, nullable: true })
  name?: string;

}

export class ChildEntity {
  @ManyToOne({
    entity: () => ParentEntity,
    inversedBy: e => e.children,
    joinColumn: 'parentEntityKey',
  })
  parentEntity!: ParentEntity;

  @PrimaryKey()
  id!: number;

  @Property({ length: 191 })
  parentEntityKey!: string;

  @Property({ length: 191, nullable: true })
  name?: string;
}
```

If we try to select all ChildEntity objects which belong to a given ParentEntity, then MikroORM does this
```ts
// try to get all child entities that belong to a given parent entity
this.childEntityRepo.find({ parentEntity: { name: 'entity-name' }});

// generates this SQL
select `e0`.* from `child_entity` as `e0` left join `parent_entity` as `e1` on `e0`.`parentEntityKey` = `e1`.`id` where `e1`.`name` = 'entity-name'
```

It also happens the other way around, e.g.
```ts
// get all ParentEntity which have children with a certain name
this.parentEntityRepo.find({ children: { name: 'entity-name' }});

// generates
select `e0`.* from `parent_entity` as `e0` left join `child_entity` as `e1` on `e0`.`id` = `e1`.`parentEntityKey` where `e1`.`name` = 'entity-name'
```

**Expected behavior**
Since I set the columns myself, I would have expected it to try this instead
```ts
select `e0`.* from `child_entity` as `e0` left join `parent_entity` as `e1` on `e0`.`parentEntityKey` = `e1`.`key` where `e1`.`name` = 'entity-name'
```

I tried using the `inverseJoinColumn`, `referenceColumnName`, and `fieldName` attributes on both the OneToMany() and ManyToOne() sides but the join condition generated by MikroORM stayed the same.

Is there something I need to change in my code, or is this a bug with MikroORM?

**Versions**

| Dependency | Version |
| - | - |
| node |  12.16.1 |
| nestjs |  7 |
| typescript | 3.9.3 |
| mikro-orm | 3.6.14 |
| mysql2 | 2.1.0 |

**Comment by B4nan** [comment]
There are few issues with your attempt, but the bottom line is that this is not really supported. ManyToOne always needs to point to the primary key of the inverse entity. Also you can't have 2 properties on your entity that represent same column (`parentEntity` vs `parentEntityKey` are in fact one column but you are modelling it as two).

If you are not planning to use ORM's schema generator, you could hack it this way:

```typescript
@Entity()
export class ChildEntity {

  @ManyToOne({
    entity: () => ParentEntity,
    inversedBy: 'children',
    joinColumn: 'parentEntityKey',
    referenceColumnName: 'key',
  } as any) // need to cast as any to be able to set `referenceColumnName`
  parentEntity!: ParentEntity;

  @PrimaryKey()
  id!: number;

  @Property({ persist: false })
  get parentEntityKey(): string {
    return this.parentEntity.key;
  }

  @Property({ length: 191, nullable: true })
  name?: string;

}
```

No need to do anything on the inverse (1:M) side (you can drop the `joinColumn` there).

**Comment by epechuzal** [comment]
Thank you, that works! Is this a 'safe' workaround to use? Specifically, do we risk this logic changing/breaking one day in the future?

Can you provide advice on how to select entities with conditions on the relation?

e.g. if I do something like this
```ts
// imagine parentEntity is an instance of ParentEntity
this.childEntityRepo.find({ parentEntity: parentEntity });
```
then mikro-orm selects with tons of conditions like
```sql
select `e0`.* from `child_entity` as `e0` left join `parent_entity` as `e1` on `e0`.`parentEntityKey` = `e1`.`key` where `e1`.`id` = 1 and `e1`.`name` = 'parent-entity-name' and .... 
// all attributes of parentEntity are in the where clause
```
but mikro-orm knows the join columns and the primary keys, so I thought it would know to do a simpler where clause with just the primary key of parentEntity
```sql
select `e0`.* from `child_entity` as `e0` left join `parent_entity` as `e1` on `e0`.`parentEntityKey` = `e1`.`key` where `e1`.`id` = 1
```
In the end as a workaround I just do this, but it feels less 'natural'
```ts
this.childEntityRepo.find({ parentEntity: { id: 1 } });
```

I am sure that I am doing something incorrect though, can you advise?

**Comment by B4nan** [comment]
Yeah I guess this should be safe, will probably expose the `referenceColumnName` option in v4. 

> then mikro-orm selects with tons of conditions like

Are you sure that `parentEntity` is an actual instance of the entity? This looks like you have just a POJO there. 

You can also do `this.childEntityRepo.find({ parentEntity: parentEntity.id });`.

**Comment by epechuzal** [comment]
I think you're right, it must have not been an instance of the class because I can't reproduce the behaviour any more.

I do need a workaround in cases where the join column is not the primary key though, like
```
const parent = await this.parentEntityRepo.findOne({ name: 'parent-name' });
// so now parent is instance of ParentEntity...

await this.childEntityRepo.find({ parentEntity: parent });
// generates incorrectly: select `e0`.* from `child_entity` as `e0` where `e0`.`parentEntityKey` = 1

await this.childEntityRepo.find({ parentEntity: parent.id });
// generates incorrectly: select `e0`.* from `child_entity` as `e0` where `e0`.`parentEntityKey` = 1

// here we have to cast the `find` arguments as FilterQuery<T> otherwise typescript gets upset
await this.childEntityRepo.find({ parentEntity: parent.key } as FilterQuery<ChildEntity>);
// generates correctly: select `e0`.* from `child_entity` as `e0` where `e0`.`parentEntityKey` = 'parent-key'
```

Is this correct? (or at least as close to correct as I can get, given the nature of these db entities)

**Comment by B4nan** [comment]
Yes, that is correct, if you pass entity somewhere, it will be always narrowed to its primary key. Again, you are kinda hacking the ORM :] that's why you need the type cast and why the first approach don't work out of box.

**Comment by rhyek** [comment]
> ```ts
>   @Property({ persist: false })
>   get parentEntityKey(): string {
>     return this.parentEntity.key;
>   }
> ```

@B4nan 's workaround worked perfectly on `5.3.1`, but had to change that part of it to the following (adapting to the example):
```ts
@Property({ persist: false })
get parentEntityKey(): string {
  return this.parentEntity.id; // for some reason .id is being populated for me instead of .key. 
}
```

**Comment by samarjit** [comment]
I was going to vote for this issue to be fixed, but looks like it is mostly fixed in 6.0.0-rc.3. Atleast the select query part is looking good. 

In my case I used EntitySchema and referenceColumnName works. Until 5.9.x it did not work.
There is a minor issue left in schema generation where I am seeing `parentEntityKey integer not null` is getting generated.

```
create table `parent_entity` (`id` integer not null primary key autoincrement, `key` text not null, `name` text null);

create table `child_entity` (`id` integer not null primary key autoincrement, `parentEntityKey` integer not null, `name` text null, constraint `child_entity_parentEntityKey_foreign` foreign key(`parentEntityKey`) references `parent_entity`(`key`) on update cascade);
create index `child_entity_parentEntityKey_index` on `child_entity` (`parentEntityKey`);
```

**Comment by B4nan** [comment]
The issue is still valid, it's just less apparent with the joined strategy, which is the new default in v6. The problem here is not just the join, its rather the general support for such FKs - it won't be possible to lazy load this, as the mechanism is completely dependent on the target entity PK.

I do have a plan to support this in 6.x completely.

**Comment by B4nan** [comment]
No, it's not in progress, since I haven't even started just yet.

**Comment by demetris-manikas** [comment]
Hi. 
I am in need of this feature as well.
@B4nan I believe that with some guidance I will be able to file a PR for this.
Interested?

**Comment by B4nan** [comment]
Depends on which part you mean. The fix I have on my mind is rather a feature, native support for FKs that point to a non PK column. For that to work quite a lot of internals need to be adjusted, namely identity map keys are now always PKs, and the select-in strategy population depends on those matching the FKs used in relations. I don't think that's something you could do yourself, not without me describing it to the detail (and I cant do that without experimenting, its just an idea now). But if you see some intermediate problem you could fix yourself to make this better, I'll be happy to accept a PR.

**Comment by kartikrao** [comment]
Having the same issue, will this be fixed ?

**Comment by B4nan** [comment]
This will be finally resolved in v7 via #7099. The long wait is over.
